closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -652,3 +653,341 @@ JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_GCMDecrypt
+    (*env)->ReleasePrimitiveArrayCritical(env, q, qNative, 0);
+    (*env)->ReleasePrimitiveArrayCritical(env, dp, dpNative, 0);
+    (*env)->ReleasePrimitiveArrayCritical(env, dq, dqNative, 0);
+    (*env)->ReleasePrimitiveArrayCritical(env, qinv, qinvNative, 0);
+    
+    if (ret == 0)
+        return -1;
+
+    return (jlong)privateRSACrtKey;
+}
+
+/* Free RSA Public/Private Key
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    destroyRSAKey
+ * Signature: (J)V
+ */
+JNIEXPORT void JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_destroyRSAKey
+  (JNIEnv *env, jclass obj, jlong rsaKey) {
+    RSA_free((RSA*)rsaKey);
+}
+
+/* RSAEP Cryptographic Primitive, RSA Public Key operation
+ * Returns -1 on error
+ *
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    RSAEP
+ * Signature: ([BI[BJ)I
+ */
+JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_RSAEP
+  (JNIEnv *env, jclass obj, jbyteArray k, jint kLen, jbyteArray m, jlong publicRSAKey) {
+
+    unsigned char* kNative;
+    unsigned char* mNative;
+
+    kNative = (unsigned char*)((*env)->GetPrimitiveArrayCritical(env, k, 0));
+    if (kNative == NULL) {
+        return -1;
+    }
+
+    mNative = (unsigned char*)((*env)->GetPrimitiveArrayCritical(env, m, 0));
+    if (mNative == NULL) {
+        (*env)->ReleasePrimitiveArrayCritical(env, k, kNative,  0);
+        return -1;
+    }
+
+    RSA* rsaKey = (RSA*)publicRSAKey;
+
+    // OSSL_RSA_public_decrypt returns -1 on error
+    int msg_len = RSA_public_decrypt(kLen, kNative, mNative, rsaKey, RSA_NO_PADDING);
+
+    (*env)->ReleasePrimitiveArrayCritical(env, k, kNative,  0);
+    (*env)->ReleasePrimitiveArrayCritical(env, m, mNative,  0);
+    return msg_len;
+}
+
+/* RSADP Cryptographic Primitive, RSA Private Key operation
+ * Returns -1 on error
+ * The param verify is -1 for 'no verify', otherwise it is size of m (with verify)
+ *
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    RSADP
+ * Signature: ([BI[BIJ)I
+ */
+JNIEXPORT jint JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_RSADP
+  (JNIEnv *env, jclass obj, jbyteArray k, jint kLen, jbyteArray m, jint verify, jlong privateRSAKey) {
+
+    unsigned char* kNative;
+    unsigned char* mNative;
+
+    kNative = (unsigned char*)((*env)->GetPrimitiveArrayCritical(env, k, 0));
+    if (kNative == NULL) {
+        return -1;
+    }
+
+    mNative = (unsigned char*)((*env)->GetPrimitiveArrayCritical(env, m, 0));
+    if (mNative == NULL) {
+        (*env)->ReleasePrimitiveArrayCritical(env, k, kNative,  0);
+        return -1;
+    }
+
+    RSA* rsaKey = (RSA*)privateRSAKey;
+
+    // OSSL_RSA_private_encrypt returns -1 on error
+    int msg_len = RSA_private_encrypt(kLen, kNative, mNative, rsaKey, RSA_NO_PADDING);
+
+    if (verify != -1 && msg_len != -1) {
+        if (verify == kLen) {
+            unsigned char* k2 = malloc(kLen * (sizeof(unsigned char)));
+            if (k2 != NULL){
```

Comment:
Please fix the style as I see has been done elsewhere, and add a space between `){`